[bitnami/redis] Fix redis sentinel timeout

nobiit opened the pull request:
<!--
 Before you open the request please review the following guidelines and tips to help it be more easily integrated:

 - Describe the scope of your change - i.e. what the change does.
 - Describe any known limitations with your change.
 - Please run any tests or examples that can exercise your modified code.

 Thank you for contributing! We will try to test and integrate the change as soon as we can, but be aware we have many GitHub repositories to manage and can't immediately respond to every request. There is no need to bump or check in on a pull request (it will clutter the discussion of the request).

 Also don't be worried if the request is closed or not integrated sometimes the priorities of Bitnami might not match the priorities of the pull request. Don't fret, the open source community thrives on forks and GitHub makes it easy to keep your changes in a forked repo.
 -->

### Description of the change
Fixed redis sentinel waiting until timeout and exiting
<!-- Describe the scope of your change - i.e. what the change does. -->

### Benefits
Some CNI will keep connection to service until 1 pods are available
<!-- What benefits will be realized by the code change? -->

### Possible drawbacks
None
<!-- Describe any known limitations with your change -->

### Applicable issues
On some CNI (Cilium) it will keep the connection to the Service until timeout, it makes redis sentinel boot longer than usual and is killed by kubernetes
<!-- Enter any applicable Issues here (You can reference an issue using #) -->
  <!-- - fixes # -->

### Additional information
On 1 test machine (Created with kind)
```
$ kubectl logs redis-node-0 sentinel 
 02:43:22.99 INFO  ==> about to run the command: REDISCLI_AUTH=$REDIS_PASSWORD redis-cli -h redis.default.svc.cluster.local -p 26379 sentinel get-master-addr-by-name mymaster
Could not connect to Redis at redis.default.svc.cluster.local:26379: Connection refused
1:X 15 Jul 2022 02:43:25.074 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
1:X 15 Jul 2022 02:43:25.074 # Redis version=7.0.2, bits=64, commit=00000000, modified=0, pid=1, just started
1:X 15 Jul 2022 02:43:25.074 # Configuration loaded
1:X 15 Jul 2022 02:43:25.075 * monotonic clock: POSIX clock_gettime
1:X 15 Jul 2022 02:43:25.076 * Running mode=sentinel, port=26379.
1:X 15 Jul 2022 02:43:25.076 # Sentinel ID is 2a09ba7abbb41ee71e79087310d75f9809c3c815
1:X 15 Jul 2022 02:43:25.076 # +monitor master mymaster redis-node-0.redis-headless.default.svc.cluster.local 6379 quorum 2
```

On my machine (using CNI Cilium)
```
$ kubectl logs -f redis-node-0 sentinel 
 08:08:14.43 INFO  ==> Found previous master redis-node-0.redis-headless.db.svc.cluster.local in /opt/bitnami/redis-sentinel/etc/sentinel.conf
 08:08:14.43 INFO  ==> about to run the command: REDISCLI_AUTH=$REDIS_PASSWORD redis-cli -h redis.db.svc.cluster.local -p 26379 sentinel get-master-addr-by-name mymaster
$ kubectl describe -n db pods redis-node-0 
Events:
  Type     Reason     Age                  From               Message
  ----     ------     ----                 ----               -------
  Normal   Scheduled  12m                  default-scheduler  Successfully assigned db/redis-node-0 to k8s-01
  Normal   Pulled     11m                  kubelet            Container image "[docker.io/bitnami/bitnami-shell:11-debian-11-r14]" already present on machine
  Normal   Created    11m                  kubelet            Created container volume-permissions
  Normal   Started    11m                  kubelet            Started container volume-permissions
  Normal   Pulled     11m                  kubelet            Container image "[docker.io/bitnami/redis:7.0.3-debian-11-r0]" already present on machine
  Normal   Created    11m                  kubelet            Created container redis
  Normal   Started    11m                  kubelet            Started container redis
  Normal   Pulled     11m                  kubelet            Container image "[docker.io/bitnami/redis-sentinel:7.0.2-debian-11-r9]" already present on machine
  Normal   Created    11m                  kubelet            Created container sentinel
  Normal   Started    11m                  kubelet            Started container sentinel
  Warning  Unhealthy  10m (x8 over 11m)    kubelet            Startup probe failed: dial tcp 10.192.0.122:6379: connect: connection refused
  Warning  Unhealthy  112s (x52 over 11m)  kubelet            Startup probe failed: dial tcp 10.192.0.122:26379: connect: connection refused
```
<!-- If there's anything else that's important and relevant to your pull request, mention that information here.-->

### Checklist

<!-- [Place an '[X]' (no spaces) in all applicable fields. Please remove unrelated fields.] -->

- [x] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/). This is *not necessary* when the changes only affect README.md files.
- [x] Variables are documented in the values.yaml and added to the `README.md` using [readme-generator-for-helm](https://github.com/bitnami-labs/readme-generator-for-helm)
- [x] Title of the pull request follows this pattern [bitnami/<name_of_the_chart>] Descriptive title
- [x] All commits signed off and in agreement of [Developer Certificate of Origin (DCO)](https://github.com/bitnami/charts/blob/master/CONTRIBUTING.md#sign-your-work)

nobiit commented:
Please accept this PR if this PR is accepted
https://github.com/bitnami/bitnami-docker-redis-sentinel/pull/51

fmulero commented:
Thanks @nobiit, I'll pass this PR to one of our engineers to follow up, but there is something it looks weird to me, we are not providing `netcat` in our images. How did you test it?

nobiit commented:
Hey @fmulero , I created 1 PR to docker redis to add netcat (https://github.com/bitnami/bitnami-docker-redis-sentinel/pull/51). I can't find any other way in the current image

fmulero commented:
Please don't do that, `netcat` is the swiss army knife to open undesired connections. Please have a look to [libnet.sh](https://github.com/bitnami/bitnami-docker-redis-sentinel/blob/974d4456c45e824a4eb0ef1f94d5d9361313d1e3/7.0/debian-11/prebuildfs/opt/bitnami/scripts/libnet.sh) or use  something like:
```console
$ timeout 1 bash -c 'cat < /dev/null > /dev/tcp/google.com/80'
```

nobiit commented:
Thanks @fmulero  for the suggestion, I've been looking for such a command and can't find it.
I will update the PR again tomorrow

nobiit commented:
Hey @fmulero, I updated the PR and tested again in my environment. It works pretty well, thank you
Please review the PR, or wait for someone else to have the same problem if you have any doubts

```
$ kubectl logs -f -n db redis-node-0 sentinel 
 02:56:34.37 INFO  ==> Found previous master redis-node-0.redis-headless.db.svc.cluster.local in /opt/bitnami/redis-sentinel/etc/sentinel.conf
 02:56:35.38 WARN  ==> Timeout when connecting to redis.db.svc.cluster.local, will use redis-headless.db.svc.cluster.local
 02:56:35.38 INFO  ==> about to run the command: REDISCLI_AUTH=$REDIS_PASSWORD redis-cli -h redis-headless.db.svc.cluster.local -p 26379 sentinel get-master-addr-by-name mymaster
Could not connect to Redis at redis-headless.db.svc.cluster.local:26379: Connection refused
1:X 16 Jul 2022 02:56:35.480 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
1:X 16 Jul 2022 02:56:35.480 # Redis version=7.0.2, bits=64, commit=00000000, modified=0, pid=1, just started
1:X 16 Jul 2022 02:56:35.480 # Configuration loaded
1:X 16 Jul 2022 02:56:35.481 * monotonic clock: POSIX clock_gettime
1:X 16 Jul 2022 02:56:35.481 * Running mode=sentinel, port=26379.
1:X 16 Jul 2022 02:56:35.482 # Sentinel ID is 2a09ba7abbb41ee71e79087310d75f9809c3c815
1:X 16 Jul 2022 02:56:35.482 # +monitor master mymaster redis-node-0.redis-headless.db.svc.cluster.local 6379 quorum 2
```

```
$ kubectl logs -f -n db redis-node-1 sentinel 
 03:01:01.17 INFO  ==> about to run the command: REDISCLI_AUTH=$REDIS_PASSWORD redis-cli -h redis.db.svc.cluster.local -p 26379 sentinel get-master-addr-by-name mymaster
redis-node-0.redis-headless.db.svc.cluster.local
6379
 03:01:01.18 INFO  ==> about to run the command: REDISCLI_AUTH=$REDIS_PASSWORD redis-cli -h redis.db.svc.cluster.local -p 26379 sentinel get-master-addr-by-name mymaster
 03:01:01.18 INFO  ==> printing REDIS_SENTINEL_INFO=(redis-node-0.redis-headless.db.svc.cluster.local,6379)
1:X 16 Jul 2022 03:01:01.270 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
1:X 16 Jul 2022 03:01:01.270 # Redis version=7.0.2, bits=64, commit=00000000, modified=0, pid=1, just started
1:X 16 Jul 2022 03:01:01.270 # Configuration loaded
1:X 16 Jul 2022 03:01:01.270 * monotonic clock: POSIX clock_gettime
1:X 16 Jul 2022 03:01:01.271 * Running mode=sentinel, port=26379.
1:X 16 Jul 2022 03:01:01.272 # Sentinel ID is 33535e4e17bf8f9f9ff9ce8f9ddf609e558ff4f2
1:X 16 Jul 2022 03:01:01.272 # +monitor master mymaster redis-node-0.redis-headless.db.svc.cluster.local 6379 quorum 2
1:X 16 Jul 2022 03:01:01.274 * +slave slave redis-node-1.redis-headless.db.svc.cluster.local:6379 redis-node-1.redis-headless.db.svc.cluster.local 6379 @ mymaster redis-node-0.redis-headless.db.svc.cluster.local 6379
1:X 16 Jul 2022 03:01:01.367 * Sentinel new configuration saved on disk
```

```
$ kubectl logs -f -n db redis-node-0 redis 
 02:56:33.89 INFO  ==> about to run the command: REDISCLI_AUTH=$REDIS_PASSWORD timeout 220 redis-cli -h redis.db.svc.cluster.local -p 26379 sentinel get-master-addr-by-name mymaster
 03:00:13.89 INFO  ==> Found previous master redis-node-0.redis-headless.db.svc.cluster.local:6379 in /opt/bitnami/redis-sentinel/etc/sentinel.conf
 03:00:13.90 INFO  ==> Configuring the node as master
1:C 16 Jul 2022 03:00:13.910 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
1:C 16 Jul 2022 03:00:13.910 # Redis version=7.0.3, bits=64, commit=00000000, modified=0, pid=1, just started
1:C 16 Jul 2022 03:00:13.910 # Configuration loaded
1:M 16 Jul 2022 03:00:13.910 * monotonic clock: POSIX clock_gettime
1:M 16 Jul 2022 03:00:13.911 * Running mode=standalone, port=6379.
1:M 16 Jul 2022 03:00:13.911 # Server initialized
1:M 16 Jul 2022 03:00:13.916 * Reading RDB base file on AOF loading...
1:M 16 Jul 2022 03:00:13.916 * Loading RDB produced by version 7.0.3
1:M 16 Jul 2022 03:00:13.916 * RDB age 87889 seconds
1:M 16 Jul 2022 03:00:13.916 * RDB memory usage when created 0.82 Mb
1:M 16 Jul 2022 03:00:13.916 * RDB is base AOF
1:M 16 Jul 2022 03:00:13.916 * Done loading RDB, keys loaded: 0, keys expired: 0.
1:M 16 Jul 2022 03:00:13.916 * DB loaded from base file appendonly.aof.1.base.rdb: 0.001 seconds
1:M 16 Jul 2022 03:00:13.916 * DB loaded from append only file: 0.001 seconds
1:M 16 Jul 2022 03:00:13.916 * Opening AOF incr file appendonly.aof.1.incr.aof on server start
1:M 16 Jul 2022 03:00:13.916 * Ready to accept connections
1:M 16 Jul 2022 03:01:00.773 * Replica redis-node-1.redis-headless.db.svc.cluster.local:6379 asks for synchronization
1:M 16 Jul 2022 03:01:00.773 * Full resync requested by replica redis-node-1.redis-headless.db.svc.cluster.local:6379
1:M 16 Jul 2022 03:01:00.773 * Replication backlog created, my new replication IDs are '26c60dedd154473cc80e75ee6a3dd125bc28374f' and '0000000000000000000000000000000000000000'
1:M 16 Jul 2022 03:01:00.773 * Delay next BGSAVE for diskless SYNC
1:M 16 Jul 2022 03:01:05.006 * Starting BGSAVE for SYNC with target: replicas sockets
1:M 16 Jul 2022 03:01:05.006 * Background RDB transfer started by pid 243
243:C 16 Jul 2022 03:01:05.008 * Fork CoW for RDB: current 0 MB, peak 0 MB, average 0 MB
1:M 16 Jul 2022 03:01:05.008 # Diskless rdb transfer, done reading from pipe, 1 replicas still up.
1:M 16 Jul 2022 03:01:05.107 * Background RDB transfer terminated with success
1:M 16 Jul 2022 03:01:05.107 * Streamed RDB transfer with replica redis-node-1.redis-headless.db.svc.cluster.local:6379 succeeded (socket). Waiting for REPLCONF ACK from slave to enable streaming
1:M 16 Jul 2022 03:01:05.107 * Synchronization with replica redis-node-1.redis-headless.db.svc.cluster.local:6379 succeeded
```

nobiit commented:
/assign @fmulero

bufke commented:
A workaround until this is fixed is to set sentinel.getMasterTimeout to a lower value such as 60. The pod will hang for that value before starting. The default is 220 which results in the pod restarting due to failing readiness checks. I suppose alternatively one could set the check delay to 220, but that's a long time to wait for the pod to start.

fmulero commented on bitnami/redis/templates/scripts-configmap.yaml:
```diff
@@ -64,10 +64,15 @@ data:
 
     SENTINEL_SERVICE_PORT=$(get_port "{{ include "common.names.fullname" . }}" "TCP_SENTINEL")
     validate_quorum() {
+        redis_service=${REDIS_SERVICE}
+        if ! timeout 1 bash -c "cat < /dev/null > /dev/tcp/${REDIS_SERVICE}/${SENTINEL_SERVICE_PORT}"; then
+            warn "Timeout when connecting to ${REDIS_SERVICE}, will use ${HEADLESS_SERVICE}"
+            redis_service=${HEADLESS_SERVICE}
+        fi
```
`$REDIS_SERVICE` could be an external master (`$REDIS_EXTERNAL_MASTER_HOST`) in that case doesn't make sense to use the headless service.

I think we could use directly the headless service in this function unless we set the external master host. Here is my proposal:

```suggestion
        redis_service=${REDIS_EXTERNAL_MASTER_HOST:-$HEADLESS_SERVICE}
```
I think there is no need to use the `timeout` validation because the headless service will publish the non ready endpoints due to its configuration.

Could you give it a try? I'll open an internal question to ensure this is correct

nobiit replied:
For the master (replicas 0) it is always unavailable (because it is starting up). Therefore, it is advisable to start via headless to get a clean reused connection instead of pointless waiting.
For slaves (or replicas greater than 0), it should only start when it finds an available connection through the service, not the headless service. This will help the redis cluster to only accept new members when the cluster is up and running

fmulero replied:
Sorry @nobiit for my very late response

I think we had a little misunderstood here (or my head is a little bit messy, that's another option). I am going to try to do a recap:

#### Current Situation

I will be very schematic and I will assume `redis` as release name:
* Pods are managed by a `statefulset`.
* redis service will use all pods from created by the `statefulset`.
* redis-headless service is a headless and it won't wait till readiness tests to expose the endpoints. It uses the same pods as the redis service.
* Each pod has 2 containers: redis and sentinel. 
* Redis container in first pod (`redis-node-0`) is configured to be master.  AFAIK It doesn't have connections to redis service.
* Sentinel container in first pod (`redis-node-0`) tries to connect to redis service to get master info. According to your logs, in this point the release is getting stuck, the statefulset doesn't create the following pod.

#### Your Solution

Change `validate_quorum` function used by redis 'replica' containers  (redis-node-1, redis-node-2, etc) during startup.


I don't fully understand how your solution fix the problem (probably I am missing something). I'll try to setup a cluster with Cillium, but I'll need some time

fmulero replied:
Sorry I have no chance to test it.

@nobiit I don't fully understand how changing the `validate_quorum` function used by the replica containers (redis-node-1, redis-node-2, etc) fixes an issue that is appearing in redis-node-0. I think I am missing something

fmulero reviewed (commented):
Hi @nobiit  I've just open an internal question to ensure my assumptions are correct but Could you give a try to my suggestion?
